src/writer/update-proposal.ts

```diff
@@ -0,0 +1,113 @@
+// because we assume that those checks were already done during the proposal creation
+export function getSpaceUpdateError({ type, space }): string | undefined {
+  const { voting = {} } = space;
+
+  if (voting.type && type !== voting.type) return 'space voting type mismatch';
+
+  return undefined;
+}
+
+export function isScamDetected({ address, name, body }): boolean {
+  const proposalNameLC = name.toLowerCase();
+  const proposalBodyLC = body.toLowerCase();
+  const addressLC = address.toLowerCase();
+
+  return (
+    flaggedAddresses.includes(addressLC) ||
+    flaggedProposalBodyKeywords.some(keyword => proposalBodyLC.includes(keyword)) ||
+    flaggedProposalTitleKeywords.some(keyword => proposalNameLC.includes(keyword))
+  );
+}
+
+export async function verify(body): Promise<any> {
+  const msg = jsonParse(body.msg);
+  const timestampNow = Math.floor(Date.now() / 1e3);
+
+  const schemaIsValid = snapshot.utils.validateSchema(snapshot.schemas.updateProposal, msg.payload);
+  if (!schemaIsValid) {
+    log.warn('[writer] Wrong proposal format', schemaIsValid);
+    return Promise.reject('wrong proposal format');
+  }
+
+  const proposal = await getProposal(msg.space, msg.payload.proposal);
+  if (!proposal) return Promise.reject('unknown proposal');
+  if (proposal.start < timestampNow) return Promise.reject('proposal already started');
+
+  const isChoicesValid = validateChoices({
+    type: msg.payload.type || proposal.type,
+    choices: msg.payload.choices.length > 0 ? msg.payload.choices : proposal.choices
+  });
+  if (!isChoicesValid) {
+    return Promise.reject(`wrong choices for "${msg.payload.type || proposal.type}" type voting`);
+  }
+
+  if (proposal.author !== body.address) return Promise.reject('Not the author');
+
+  const space = await getSpace(msg.space);
+  space.id = msg.space;
+
+  const spaceUpdateError = getSpaceUpdateError({
+    type: msg.payload.type || proposal.type,
+    space
+  });
+  if (spaceUpdateError) return Promise.reject(spaceUpdateError);
+
+  const hasScam = isScamDetected({
+    address: body.address,
+    name: msg.payload.name || proposal.title,
+    body: msg.payload.body || proposal.body
+  });
+  if (hasScam) return Promise.reject('scam proposal detected, contact support');
+
+  return Promise.resolve(proposal);
+}
+
+export async function action(body, ipfs, receipt, id, context): Promise<void> {
+  const originalProposal = context;
+  const msg = jsonParse(body.msg);
+  const updated = parseInt(msg.timestamp);
+  const metadata = msg.payload.metadata || {};
+  const plugins = JSON.stringify(metadata.plugins || {});
+
+  const proposal = {
+    ipfs,
+    updated,
+    type: msg.payload.type || originalProposal.type,
+    plugins: plugins || originalProposal.plugins,
+    title: msg.payload.name || originalProposal.title,
+    body: msg.payload.body,
```

Comment:
This will assume that on every update, you always send the `body`, updated or not, else if I update only the `title`, and only send the `title` in the payload, it'll wipe out the existing body.

Reply:
UI always needs to send all those fields, because we sign our data with [signTypedData](https://docs.ethers.org/v5/api/signer/#Signer-signTypedData). The fallback to `originalProposal` is used to restrict a user from setting some fields as empty. In this particular case I have assumed that `body` could be empty, but `title`, for example, cannot be

Reply:
If UI always send the `title` field, why do you need to fallback to the original proposal ? Empty field prevention should already be done from the schema, isn't it ?

```json
"name": {
  "type": "string",
  "title": "name",
  "minLength": 1,
  "maxLength": 256
},
```

Reply:
because `Ajv` doesn't validate it.